main/lsp/ErrorReporter.cc

```diff
@@ -0,0 +1,128 @@
+#include "main/lsp/ErrorReporter.h"
+#include "core/lsp/TypecheckEpochManager.h"
+#include "main/lsp/LSPMessage.h"
+#include "main/lsp/LSPOutput.h"
+#include "main/lsp/json_types.h"
+
+namespace sorbet::realmain::lsp {
+using namespace std;
+ErrorReporter::ErrorReporter(shared_ptr<const LSPConfiguration> config) : config(move(config)) {}
+
+// Used for unit tests
+const std::vector<ErrorStatus> &ErrorReporter::getFileErrorStatuses() const {
+    return fileErrorStatuses;
+};
+
+// Used for unit tests
+const UnorderedMap<core::FileRef, ErrorStatus> &ErrorReporter::getUncommittedFileErrorStatuses() const {
+    return uncommittedFileErrorStatuses;
+};
+
+void ErrorReporter::setMaxFileId(u4 id) {
+    ENFORCE(id >= fileErrorStatuses.size());
+    fileErrorStatuses.resize(id + 1, ErrorStatus{0, false});
+};
+
+void ErrorReporter::commit() {
+    for (auto &uncommitted : uncommittedFileErrorStatuses) {
+        auto fileId = uncommitted.first.id();
+        if (fileId >= fileErrorStatuses.size()) {
+            setMaxFileId(fileId);
+        }
+
+        fileErrorStatuses[fileId] = uncommitted.second;
+    }
+
+    uncommittedFileErrorStatuses.clear();
+};
+
+vector<core::FileRef> ErrorReporter::abort() {
```

Comment:
I just realized an issue with this design. I think it's also be a bug in the previous version of the code -- and it explains a bug I noticed where sometimes old errors aren't cleared! 🎉 

Consider the following scenario:

- `foo.rb` has no errors.
- A slow path edit introduces error in file `foo.rb`.
- A second slow path edit fixes error in file `foo.rb`, canceling the first.

At the end of `abort()`, `ErrorReporter` will think that the client thinks that `foo.rb` has no errors. When it's retypechecked, it has no errors, so no updated error list is sent for the file.

I think I figured out a cleaner design to handle this scenario, though. And hopefully it doesn't frustrate you.

- Get rid of `commit` and `abort`, and `uncommittedFileErrorStatuses`. All file error statuses are immediately saved into `fileErrorStatuses`.
- Add a new method `vector<core::FileRef> filesUpdatedSince(u4 epoch)`, and return all filerefs who have a sentEpoch >= epoch.
- Call `filesUpdatedSince()` in the place we call `abort()` with the `epoch` of the canceled slow path.
- Save the `epoch` of the running slow path into `UndoState` for this purpose.

In this scenario, LSPTypechecker will need to filter out new files in `filesUpdatedSince`.

Before doing that, though, we should probably add a failing test case that reveals the bug I've described.